apex/contrib/group_norm/group_norm.py

```diff
@@ -0,0 +1,209 @@
+def inductor_group_norm(x, g, w, b, eps, act=""):
+    y = F.group_norm(x, g, w, b, eps)
+    if act in ["silu", "swish"]:
+        y = F.silu(y)
+    return y
+
+
+class GroupNormNHWC(torch.autograd.Function):
+
+    @staticmethod
+    @disable  # This shouldn't be captured by TorchDynamo
+    def forward(ctx,
+                x,
+                G,
+                weight,
+                bias,
+                eps,
+                act="",
+                algo=group_norm_cuda.OnePass):
+        # sanity check
+        act = act.lower()
+        assert x.is_contiguous(memory_format=torch.channels_last), \
+            "Only support NHWC layout."
+        assert weight.numel() == x.shape[1], "Unexpected parameter count."
+        assert bias.numel() == x.shape[1], "Unexpected parameter count."
+        assert x.shape[1] % G == 0, "C % G != 0."
+        assert act in ["", "silu", "swish"], "Unsupported activation."
+
+        with_swish = (act in ["silu", "swish"])
+
+        # enqueue fprop kernel
+        y, sums = group_norm_cuda.forward(x, G, weight, bias, eps, algo,
+                                          with_swish)
+
+        # save for backward
+        ctx.save_for_backward(x, weight, bias, sums)
+        ctx.G = G
+        ctx.eps = eps
+        ctx.algo = algo
+        ctx.with_swish = with_swish
+        return y
+
+    @staticmethod
+    def backward(ctx, dy):
+        # sanity check
+        assert dy.is_contiguous(memory_format=torch.channels_last), \
+            "Only support NHWC layout."
+
+        # retrive saved info
+        x, w, b, sums = ctx.saved_tensors
+        G = ctx.G
+        eps = ctx.eps
+        algo = ctx.algo
+        with_swish = ctx.with_swish
+
+        # enqueue bprop kernel
+        dx, dw, db = group_norm_cuda.backward(dy, sums, x, G, w, b, eps, algo,
+                                              with_swish)
+
+        return dx, None, dw, db, None, None, None
+
+
+class GroupNormOnePass(GroupNormNHWC):
+
+    @staticmethod
+    @disable
+    def forward(ctx, x, G, weight, bias, eps, act=""):
+        return super(GroupNormOnePass,
+                     GroupNormOnePass).forward(ctx, x, G, weight, bias, eps,
+                                               act, group_norm_cuda.OnePass)
+
+
+class GroupNormTwoPass(GroupNormNHWC):
+
+    @staticmethod
+    @disable
+    def forward(ctx, x, G, weight, bias, eps, act=""):
+        return super(GroupNormTwoPass,
+                     GroupNormTwoPass).forward(ctx, x, G, weight, bias, eps,
+                                               act, group_norm_cuda.TwoPass)
+
+
+cuda_group_norm_nhwc = GroupNormNHWC.apply
+cuda_group_norm_nhwc_one_pass = GroupNormOnePass.apply
+cuda_group_norm_nhwc_two_pass = GroupNormTwoPass.apply
+
+
+# We do not direct inherit from torch.nn.GroupNorm since several fusers don't
+# support inheritance. Extends:
+# https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/normalization.py
+class GroupNormOpt(torch.nn.Module):
```

Comment:
QQ: what would `Opt` mean here?

Reply:
It's short for 'Optimized'. To make it clear, I removed `Opt`, so that users will only import `GroupNorm` from `apex`.